File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pre = [1] * (len(nums))
        post = [1] * (len(nums))
        prev = 1
        last = 1
        ans = []
        for i, v in enumerate(nums):
            pre[i]=prev
            post[len(nums)-i-1]=last
            
            prev = prev * v
            last = last * nums[len(nums)-i-1]
        for i in range(len(nums)):
            ans.insert(i,pre[i]*post[i])
        return ans

        

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans=new int[nums.length];
        int[] pre = new int[nums.length];
        int[] post = new int[nums.length];
        int pre_prod=1;
        int pos_prod=1;
        for(int i=0;i<nums.length;i++){
            pre[i] = pre_prod;
            pre_prod *= nums[i];
        }
        for(int i=nums.length-1;i>=0;i--){
            post[i]=pos_prod;
            pos_prod *= nums[i];
        }
        for(int i=0;i<nums.length;i++){
            ans[i]=pre[i]*post[i];
        }
        return ans;
    }
}